### ISSUE TITLE
Partitioned Aggregation Support

### ISSUE THREAD

**Issue by dario-liberman** [apache/pinot#12057]
Add support to optimise aggregation for partitioned columns.

An example where the optimisation would be beneficial is distinct count, as it is sufficient to count distinct values within each partition and then a simple sum of the counts per partition would yield the final aggregate result.

There is in fact an aggregation function today that uses this notion, [`SEGMENTPARTITIONEDDISTINCTCOUNT` ](https://docs.pinot.apache.org/configuration-reference/functions/segmentpartitioneddistinctcount), however it only handles cases where the whole partition fits in a single segment, so it is not applicable to most Pinot use cases (eg time-based).

A reasonable approach is to support the partitioned aggregation optimisation for [partitioned segment assignment](https://docs.pinot.apache.org/operators/operating-pinot/segment-assignment#partitioned-replica-group-segment-assignment) configurations, so we can effectively assume that all segments for a given partition will be co-located on the same server.   

This means that there are two key aspects that remain the same:
1. Aggregation within each segment remains the same, as a segment belongs to a single partition.
2. Broker aggregation across server results remains the same, all partitioning concerns can be resolved within the server.

The proposal is for the [`AggregationFunction`](https://github.com/apache/pinot/blob/master/pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunction.java) to have an extra stage allowing to merge results within each partition if the partitioning optimisation applies. 

There are multiple ways to go about supporting this.

New methods could look something like this:
``` 
boolean isPartitionedAggregation();
IntermediateResult extractPartitionResult(IntermediateResult intermediateResult);
```

A more involved change would have a new generic type added, and consequently a new merge function to aggregate across partitions.
``` 
boolean isPartitionAggregation();
PartitionResult extractPartitionResult(IntermediateResult intermediateResult);
PartitionResult mergePartitions(PartitionResult partitionResult1, PartitionResult partitionResult2);
FinalResult extractFinalResult(PartitionResult partitionResult);
```

An even more involved change would allow to have a different column type (tho this is likely unnecessary given `OBJECT` can be used anyways).
```
ColumnDataType getPartitionResultColumnType();
```

We can either add default methods to `AggregationFunction` (eg returning `false`) or we can have a new interface `PartitionedAggregationFunction` extending `AggregationFunction` (then where necessary check if the aggregation function object implements the new interface). 
The latter has the benefit that we could more easily introduce a new generic type to represent the partition level intermediate result.

The flow at a high level would be:
1. Aggregate blocks within segment as usual.
2. Extract intermediate aggregation result as usual.
3. Aggregate across segments as usual with `merge` function, but keeping aggregation across partitions separate.
4. Extract partition result from each aggregated partition via `extractPartitionResult`.
5. Merge across partitions via `mergePartitions`.
6. Extract the final result from merged partitioned result.

Another alternative is to consider the `IntermediateResult` to be the final phase of aggregation across partitions and instead add a new merge function for inter-segment aggregation within partitions, in this case the type within partitions is the one that is "different".
```
boolean isPartitionedAggregation();
PartitionedResult extractAggregationResult(AggregationResultHolder aggregationResultHolder);
PartitionedResult extractGroupByResult(GroupByResultHolder groupByResultHolder, int groupKey);
PartitionedResult mergeWithinPartition(PartitionedResult partitionedResult1, PartitionedResult partitionedResult2);
IntermediateResult extractPartitionResult(PartitionedResult partitionedResult);
```

In this alternative the flow at a high level would be:
1. Aggregate blocks within segment as usual.
2. Extract partitioned results via `extractAggregationResult` or `extractGroupByResult`.
3. Aggregate across segments within partition via `mergeWithinPartition`
4. Extract intermediate result from each aggregated partition via `extractPartitionResult`.
5. Merge across partitions via regular `merge`.
6. Extract the final result from intermediate result as usual.

The advantage of this alternative is that we could possibly consider the PartitionedResult to be an internal structure to the server that does not need to be serialized, similar to block level aggregation, then the intermediate result continues to be used for broker level aggregation.

In either case, the core of the changes to the aggregation logic in order to support this partitioning optimisation would be mainly in `AggregationCombineOperator`, `GroupByCombineOperator` and `AggregationResultsBlockMerger`.

Basically, the intermediate aggregation would be done across segments within each partition based on which partition the segment belongs to, partition results can then be extracted and finally merged across partitions.

**Comment by walterddr** [comment]
is this similar to the multi-stage engine partitioned group-by optimization introduced in #11266 . we can potentially unify the 2 logics if we use broker to handle this optimization yes?

**Comment by dario-liberman** [comment]
@Jackie-Jiang - I think we can use some of the building blocks of `serverReturnFinalResult`, but it still has two gaps as far as I can see.

1. We still need a way to aggregate across servers, my point is that the merge functions within partition and across partitions are different. For example when calculating distinct counts, the former merge would be a union of sets whereas the latter merge would be a sum of numbers.
2. If we only worry about the broker, we still miss on optimisation that can be done at the server level by having more efficient cross-partition aggregation (or more importantly lower memory cost within partition, such as smaller sets for distinct count). There could be an order of magnitude (or more) difference between partition count and server count (eg. 512 partitions split across 16 servers), specially with replication there could be a deployment configuration where the cluster is large but the partitions could be co-located on a small number of servers sufficient to return the result for a given query.

In order to address both points above, I don't see how we can do it without explicit partitioned support in `AggregationFunction` (or as I say above, we can have a `PartitionedAggregationFunction` sub-class and we can check in run-time if the object implements the sub-class).

@walterddr - I was not aware of the support for partitioned multi-stage query optimisations, thanks for sharing. I agree this work could indeed also benefit partitioned optimizations in multi-stage engine. However, as I say above, it is a different problem to optimize aggregations where the group-by column is partitioned vs optimising aggregations where the partitioning allows to optimise the aggregation itself (eg distinct count over a partitioned column). I argue that the aggregation function needs to be written in such a way to allow for the optimisation.

There is a way to make this more generic I guess, we could have perhaps a pseudo function representing the partition id (could be just a number from 1 to N), then ask to write the query having two clearly separate phases. For example:

```
select SUM(partitioned_agg) from (
    select DISTINCT_COUNT(partition_column) as partitioned_agg from table 
    group by PARTITION_ID(partition_column)
)
```
Then it is the user that needs to know what function to use to aggregate across partitions.
I did not propose this, because in my view the engine should figure out how to optimise partitioning, not the user. But it is an option I guess. Maybe even ask to introduce appropriate hints for both selects above.

**Comment by walterddr** [comment]
I see. i think one misconception is this deals with more than just what multi-stage is doing: e.g. 
1. not only the group-key is partitioned, thus the agg can directly return final result on each group
2. but also if the aggregate operand is by itself partitioned, the agg function can produce another form of merge-able format that's not as complex as normal IntermediateResult. 

in this sense, i agree with @Jackie-Jiang's comment above.

The only thing I want to call out is that the `boolean isPartitionedAggregation()` is confusing. b/c it doesn't clearly tell whether it meant scenario (1) or (2) for the above. 

I think a way to clearly indicate whether the result of the aggregate is "partitionResult" or "finalResult" would be more desirable (after all that's the goal of the flag is for yes?)
